File: backend/app/schemas/self_assessment.py
from typing import Optional, TYPE_CHECKING
from uuid import UUID
from pydantic import BaseModel, Field
from datetime import datetime
from .common import SelfAssessmentStatus, RatingCode, PaginatedResponse

if TYPE_CHECKING:
    pass


class SelfAssessmentBase(BaseModel):
    self_rating_code: Optional[RatingCode] = Field(
        None,
        alias="selfRatingCode",
        description="Letter grade: SS, S, A, B, C, D (6-level input scale)"
    )
    self_comment: Optional[str] = Field(
        None,
        alias="selfComment",
        max_length=5000,
        description="Employee's narrative self-assessment comment"
    )


class SelfAssessmentCreate(SelfAssessmentBase):
    """Request schema for creating a self-assessment (system auto-creates when goal approved)."""
    status: SelfAssessmentStatus = Field(
        default=SelfAssessmentStatus.DRAFT,
        description="Assessment status: draft, submitted, or approved"
    )

    model_config = {"populate_by_name": True}


class SelfAssessmentUpdate(BaseModel):
    """Request schema for updating a self-assessment."""
    self_rating_code: Optional[RatingCode] = Field(
        None,
        alias="selfRatingCode",
        description="Letter grade: SS, S, A, B, C, D (6-level input scale)"
    )
    self_comment: Optional[str] = Field(
        None,
        alias="selfComment",
        max_length=5000,
        description="Employee's narrative self-assessment comment"
    )
    rating_data: Optional[dict] = Field(
        None,
        alias="ratingData",
        description="Granular per-action ratings for コンピテンシー goals (JSONB). NULL for 業績目標."
    )

    model_config = {"populate_by_name": True}


class SelfAssessmentInDB(SelfAssessmentBase):
    id: UUID
    goal_id: UUID
    period_id: UUID
    self_rating: Optional[float] = Field(
        None,
        alias="selfRating",
        ge=0,
        le=7,
        description="Numeric rating (0.0-7.0), auto-calculated from selfRatingCode"
    )
    rating_data: Optional[dict] = Field(
        None,
        alias="ratingData",
        description="Granular per-action ratings for コンピテンシー goals (JSONB). NULL for 業績目標."
    )
    status: SelfAssessmentStatus = SelfAssessmentStatus.DRAFT
    submitted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True, "populate_by_name": True}


class SelfAssessment(SelfAssessmentInDB):
    """
    Basic self-assessment schema for API responses (list views, simple references).
    """
    goal_id: UUID = Field(..., alias="goalId")
    period_id: UUID = Field(..., alias="periodId")
    submitted_at: Optional[datetime] = Field(None, alias="submittedAt")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    model_config = {"from_attributes": True, "populate_by_name": True}


class SelfAssessmentDetail(SelfAssessmentInDB):
    """
    Detailed self-assessment schema for single item views.
    """
    goal_id: UUID = Field(..., alias="goalId")
    period_id: UUID = Field(..., alias="periodId")
    submitted_at: Optional[datetime] = Field(None, alias="submittedAt")
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")

    # Assessment state information
    is_editable: bool = Field(True, alias="isEditable", description="Whether this assessment can still be edited")
    is_overdue: bool = Field(False, alias="isOverdue", description="Whether this assessment is past the deadline")
    days_until_deadline: Optional[int] = Field(None, alias="daysUntilDeadline", description="Days remaining until assessment deadline")

    # Goal context
    goal_category: Optional[str] = Field(None, alias="goalCategory", description="Category of the goal being assessed")
    goal_status: Optional[str] = Field(None, alias="goalStatus", description="Current status of the goal")

    model_config = {"from_attributes": True, "populate_by_name": True}


class SelfAssessmentList(PaginatedResponse[SelfAssessment]):
    """Schema for paginated self-assessment list responses"""
    pass


class SubordinateAssessmentStatus(BaseModel):
    """Assessment status for a single subordinate."""
    user_id: UUID = Field(..., alias="userId")
    total_count: int = Field(..., alias="totalCount", description="Total number of assessments")
    submitted_count: int = Field(..., alias="submittedCount", description="Number of submitted/approved assessments")
    all_submitted: bool = Field(..., alias="allSubmitted", description="Whether all assessments are submitted")
    approved_count: int = Field(..., alias="approvedCount", description="Number of approved assessments")
    all_approved: bool = Field(..., alias="allApproved", description="Whether all assessments are approved (evaluation complete)")

    model_config = {"populate_by_name": True}


class SubordinatesAssessmentStatusResponse(BaseModel):
    """Response schema for subordinates assessment status endpoint."""
    items: list[SubordinateAssessmentStatus] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


# ========================================
# FORWARD REFERENCES UPDATE
# ========================================

try:
    SelfAssessmentDetail.model_rebuild()
except Exception as e:
    print(f"Warning: Could not rebuild forward references in self_assessment schemas: {e}")
    pass

File: frontend/src/api/types/self-assessment.ts
import type {
  UUID,
  RatingCode,
  SelfAssessmentStatus,
  PaginatedResponse,
} from './common';
import type { GoalResponse } from './goal';
import type { EvaluationPeriod } from './evaluation-period';
import type { UserProfileOption } from './user';

/**
 * Self Assessment type definitions
 * These types match the backend Pydantic schemas for SelfAssessment-related operations
 * @see .kiro/specs/self-assessment/frontend-typescript-types.md
 */

/**
 * Per-action ratings for competency goals
 * Structure: { competencyId: { actionIndex: RatingCode } }
 * Example: { "uuid-1": { "0": "A", "1": "B" }, "uuid-2": { "0": "S" } }
 */
export type CompetencyRatingData = Record<string, Record<string, RatingCode>>;

/**
 * Base self-assessment fields (editable by employee)
 */
export interface SelfAssessmentBase {
  /** Letter grade: SS, S, A+, A, A-, B, C, D */
  selfRatingCode?: RatingCode;
  /** Employee's narrative self-assessment comment */
  selfComment?: string;
}

/**
 * Self-assessment entity (API response)
 * @see .kiro/specs/self-assessment/api-contract.md Section 4.1
 */
export interface SelfAssessment extends SelfAssessmentBase {
  /** Unique identifier */
  id: UUID;
  /** Reference to the goal being assessed */
  goalId: UUID;
  /** Reference to the evaluation period */
  periodId: UUID;
  /** Numeric rating (0.0-7.0), auto-calculated from selfRatingCode */
  selfRating?: number;
  /** Granular per-action ratings for コンピテンシー goals. NULL for 業績目標. */
  ratingData?: CompetencyRatingData;
  /** Current status: draft, submitted, or approved */
  status: SelfAssessmentStatus;
  /** Timestamp when assessment was submitted */
  submittedAt?: string;
  /** Record creation timestamp */
  createdAt: string;
  /** Last modification timestamp */
  updatedAt: string;
}

/**
 * Detailed self-assessment with additional context
 * Used for single item views (GET /self-assessments/:id)
 * @see .kiro/specs/self-assessment/api-contract.md Section 4.4
 *
 * NOTE: Goal title should be extracted using getGoalTitle() helper:
 * - Performance goals: use goal.title
 * - Competency goals: use competency names or fallback
 */
export interface SelfAssessmentDetail extends SelfAssessment {
  // Assessment state information
  /** Whether this assessment can still be edited */
  isEditable: boolean;
  /** Whether this assessment is past the deadline */
  isOverdue: boolean;
  /** Days remaining until assessment deadline */
  daysUntilDeadline?: number;

  // Goal context (convenience fields, extracted from goal relationship)
  /** Category of the goal being assessed (e.g., '業績目標', 'コンピテンシー') */
  goalCategory?: string;
  /** Current status of the goal */
  goalStatus?: string;

  // Embedded relationships
  /** The goal being assessed - use getGoalTitle() helper to extract title */
  goal?: GoalResponse;
  /** The evaluation period this assessment belongs to */
  evaluationPeriod?: EvaluationPeriod;
  /** The employee who owns this assessment */
  employee?: UserProfileOption;
}

/**
 * Self-assessment with embedded goal information (for list views)
 * Goal title should be extracted using getGoalTitle() helper
 */
export interface SelfAssessmentWithGoal extends SelfAssessment {
  goal: GoalResponse;
}

/**
 * Request body for updating a self-assessment
 * @see .kiro/specs/self-assessment/api-contract.md Section 4.5
 */
export interface SelfAssessmentUpdate {
  /** Letter grade: SS, S, A+, A, A-, B, C, D */
  selfRatingCode?: RatingCode;
  /** Employee's narrative self-assessment comment */
  selfComment?: string;
  /** Granular per-action ratings for コンピテンシー goals. NULL for 業績目標. */
  ratingData?: CompetencyRatingData;
}

/**
 * Paginated list of self-assessments
 * @see .kiro/specs/self-assessment/api-contract.md Section 4.1
 */
export interface SelfAssessmentList extends PaginatedResponse<SelfAssessment> {}

/**
 * Query parameters for fetching self-assessments
 */
export interface SelfAssessmentQueryParams {
  periodId?: UUID;
  userId?: UUID;
  status?: SelfAssessmentStatus;
  goalCategory?: string;
  page?: number;
  limit?: number;
}

/**
 * Assessment status for a single subordinate
 * Used by supervisors to quickly see submission status of all subordinates
 */
export interface SubordinateAssessmentStatus {
  /** Subordinate user ID */
  userId: UUID;
  /** Total number of assessments */
  totalCount: number;
  /** Number of submitted/approved assessments */
  submittedCount: number;
  /** Whether all assessments are submitted */
  allSubmitted: boolean;
  /** Number of approved assessments (evaluation complete) */
  approvedCount: number;
  /** Whether all assessments are approved (evaluation complete) */
  allApproved: boolean;
}

/**
 * Response for subordinates assessment status endpoint
 */
export interface SubordinatesAssessmentStatusResponse {
  items: SubordinateAssessmentStatus[];
}

